README.md
# dump1090-fa Debian/Raspbian packages

This is a fork of [dump1090-fa](https://github.com/flightaware/dump1090)
customized for use the --net-only HTTP server pulling data from existing
data sources. e.g. a Pi running a local ADS-B/ModeS receiver.
Intended to be run in a Docker Container

It is designed to build from the Debian.stable package. (Aug 2018)

### Modifications re original dump1090-fa

References to rtl-sdr and bladeRF are removed
--> It is not intended to collect data from a SDR receiver.
added daemon init and support for external configuration sources

### Instrument Panel
Instrument Panel for the selected aircraft, switch between panel and sheet view.
javascript draws on HTML5 Canvas.

### Screenshot and Video of the Instrument Panel

[Using the dump1090-fa with instrument panel (Youtube](https://www.youtube.com/watch?v=mQ3SeIqvK6E)
[![dump1090 instrument panel](https://raw.githubusercontent.com/bm98/dump1090/master/img/dump1090-fa-Mod1.png)](https://www.youtube.com/watch?v=mQ3SeIqvK6E)

[Panel view](https://raw.githubusercontent.com/bm98/dump1090/master/img/dump1090-fa-Mod0.png)

[Change between Panel and Sheet](https://raw.githubusercontent.com/bm98/dump1090/master/img/dump1090-fa-Mod2.png)

[Sheet view](https://raw.githubusercontent.com/bm98/dump1090/master/img/dump1090-fa-Mod3.png)


### Using only the Instrument Panel

... Stop the http daemon:  $ sudo service lighttpd stop
 
In /usr/share/dump1090-fa/html
 
Backup the original index.html, style.css and script.js
 
Drop the following files from this public_html folder here:
index.html, style.css, script.js + all the new gp*.js
 
$ sudo cp   is usually needed  
 
Start the http daemon: $ sudo service lighttpd start
 

### Actually building it

Nothing special, just build it ("dpkg-buildpackage -b")
see also docker-dump1090-fa Git repo of this Git account

no Raspian build tested so far - may or may not work..

## Building manually

You can probably just run "make" after installing the required dependencies.
Binaries are built in the source directory; you will need to arrange to
install them (and a method for starting them) yourself.

# Change notes within 3.6.2.1

20180910
added some more null value checks - indicates mostly '- - -' when the value is null
added swithing to STD BARO (1013.25) when crossing TA (default 20000ft - QHN correction)
Config file should allow to change the TA level by defining the variable:
  DefaultTransitionAltitude = 10000   // [ft]  enter whatever you think is appropriate
added nav_qnh to infowindow

EOD


debian/changelog
dump1090-fa (3.6.3.1) UNRELEASED; urgency=low

  * merge pull from upstream
  * reverted the sheet display to the original FA version
  * Updated: aircraft database from a most recent source 
    (see README in db folder)
  * NOTE: only the html section is ready to use right now

 -- Martin Burri <github@mail.burri-web.org>  Thu, 20 Sep 2018 00:42:30 +0100

dump1090-fa (3.6.2.1) UNRELEASED; urgency=low

  * initial fork from original source
  * Prep for docker use (no Net only, RTL-SDR, BLADERF)
  * Removed: files and libs for rtl-sdr, bladeRF, libUsb dependencies
  * Added: daemon init for dump1090-fa

 -- Martin Burri <github@mail.burri-web.org>  Sat, 26 Aug 2018 18:32:30 +0100

dump1090-fa (3.6.3) stable; urgency=medium

  * Fix port 30003 output (Basestation) timestamp formatting broken in 3.6.0
  * Ignore altitudes in DF17 "airborne" positions with airGround = ground
  * SkyView layout change and proper support for extended data fields

 -- Oliver Jowett <oliver@mutability.co.uk>  Tue, 11 Sep 2018 14:52:20 +0100

dump1090-fa (3.6.2) stable; urgency=medium

  * Update SkyView flags for Libya, Serbia, Montenegro
  * Improve handling of bad callsigns in ADS-B messages
  * Update FlightAware aircraft static data export
  * Add --direct option to enable rtlsdr direct sampling mode

 -- Oliver Jowett <oliver@mutability.co.uk>  Fri, 27 Jul 2018 18:32:30 +0100

dump1090-fa (3.6.1) stable; urgency=medium

  * Fix tracking of NICbaro
  * Don't translate 7x00 to emergency status, it obscures the source of the
    data (squawk vs ADS-B v2)
  * Support for building on stretch

 -- Oliver Jowett <oliver@mutability.co.uk>  Mon, 16 Jul 2018 14:58:59 +0100

dump1090-fa (3.6.0) stable; urgency=medium

  * dump1090: implement speculative decoding for Comm-B messages, capturing
    most of the Mode S EHS data:
    BDS 1,0 (datalink capabilities); BDS 1,7 (GICB capabilities);
    BDS 2,0 (aircraft identification); BDS 3,0 (ACAS RA);
    BDS 4,0 (vertical intention); BDS 5,0 (track and turn);
    BDS 6,0 (heading and speed)
  * dump1090: implement decoding of most of DO-260A/DO-260B including
    ADS-B version, autopilot settings, NIC/Rc, NAC, SIL, GVA, etc.
  * dump1090: aircraft state tracking overhaul; track data age in milliseconds
  * dump1090: emit newly decoded fields in aircraft.json
  * Skyview: experimental display of new fields, disabled by default; to
    enable, set ExtendedData = true in config.js or add #extended to the URL
  * Skyview: Remove obsolete chartbundle WAC layer, add heli layer
  * Skyview: Remove Mapzen layer as Mapzen is shutting down
  * faup1090: Restructure emitted data to include data age and source
  * faup1090: Include newly decoded fields in reports
  * dump1090, faup1090: fix use of snprintf to correctly handle buffer overruns
  * dump1090: don't bother tracking ICAO 000000

 -- Oliver Jowett <oliver@mutability.co.uk>  Wed, 09 May 2018 22:34:24 +0100

dump1090-fa (3.5.3) stable; urgency=medium

  * Skip 3.5.2 to align with piaware versioning
  * Skyview: Update OpenLayers to fix performance problems on some browsers

 -- Oliver Jowett <oliver@mutability.co.uk>  Fri, 03 Nov 2017 22:44:40 +0000

dump1090-fa (3.5.1) stable; urgency=medium

  * Skyview: fix font mappings
  * Skyview: mark tisb_icao and tisb_other addresses as TIS-B
  * Skyview: fix unit selection on Firefox
  * Skyview: report data source correctly in popup
  * Skyview: change default color for grounded aircraft to brown
  * Skyview: initial history load speedups (Steven Davies)
  * Skyview: fix L2T/L2P icon selection
  * Makefile cleanups (Michael Tatarinov)

 -- Oliver Jowett <oliver@mutability.co.uk>  Mon, 31 Jul 2017 13:43:32 +0100

dump1090-fa (3.5.0) stable; urgency=medium

  * SkyView map UI added.
  * Only enable lighttpd config once, not on every package upgrade in case the
    user has deliberately disabled it.
  * Fix AGC gain setting.
  * Updated the FlightAware aircraft data export.
  * Minor bugfixes.

 -- Oliver Jowett <oliver@mutability.co.uk>  Mon, 03 Apr 2017 17:53:48 +0100

dump1090-fa (3.4.0) stable; urgency=medium

  * Add bladeRF support.
  * Clean up the FlightAware photos link so it works for registrations containing
    non-alphanumeric characters.

 -- Oliver Jowett <oliver@mutability.co.uk>  Thu, 09 Feb 2017 17:41:43 +0000

dump1090-fa (3.3.0) stable; urgency=medium

  * Improvements to the Mode A/C demodulator.
  * Automatically enable Mode A/C if a Beast command requesting it is received.
    (disable with --no-modeac-auto)
  * Process the radarcape position message, emit it from faup1090.
  * If view1090 is started with --modeac, request Mode A/C data from dump1090.
  * Remove the last traces of the internal webserver.
  * Updated the FlightAware aircraft data export.
  * Updated the default package-based maximum range to 360NM to match what
    piaware-support configured for sdcard installs.
  * Remove internal webserver code entirely (previously, it was just disabled
    at compile time).
  * Tweak the timing reported for Mode S and Mode A/C messages to match how
    the Beast/Radarcape does it. This affects the interval between Mode S
    and Mode A/C messages only; intervals between Mode S messages and other
    Mode S messages, or A/C and other A/C, are unchanged.

 -- Oliver Jowett <oliver@mutability.co.uk>  Sun, 15 Jan 2017 13:35:56 +0000

dump1090-fa (3.2.0) stable; urgency=medium

  * Fix a stray extra comma in port 30003 output if only GNSS altitude
    was available
  * Track CPR encoding type directly rather than inferring it from airground;
    using airground meant some position messages that should be usable were
    not used
  * Mode A/C aircraft association tuning, allows dump1090 to handle much
    higher A/C rates without chewing tons of CPU
  * Map: Decrease the maximum marker size shown
  * Map: Show additional datasource information (TIS-B etc) in table mode
  * Map: Move the datablock so it doesn't overlap the zoom control

public_html/gp-compass.js
// JavaScript Document
// Handles the GlasPanel Compass / Heading items
"use strict";

// an object carrying the geometrical positions
// based on a 1000x1000 canvas - 
// some are calculated to avoid inconsistencies
function Comp_geo()
{
	// absolute pix in the canvas
	this.R    = 208;    // Radius
	this.W    = this.R*2;  // width 
	this.H    = this.W;  // height
	this.X    = gp_geo.RollXC-this.R;   // X pos relative to Turn Center
	this.Y    = 760-this.R;  // Y pos

	// DEVELOP
//	this.X    = 0;   // X pos relative to Turn Center
//	this.Y    = gp_geo.RollYC;  // Y pos

	// relative pix in the rect
	this.XC = this.W / 2; // rel center X
	this.YC = this.H / 2; // rel center Y
	
	//  Heading relative to main
	this.HeadW = 100;  		// width
	this.HeadH = 45; 		// height MASTER for the others
	this.HeadX = this.XC-this.HeadW/2;   //  X pos
	this.HeadY = -this.HeadH; // Y pos (above compass)

	this.HeadXC = this.HeadW / 2; // rel center X
	this.HeadYC = this.HeadH / 2; // rel center Y
	
	//  NAV-Heading relative to main
	this.HeadNavW = 110;  	// width
	this.HeadNavH = 40; 	// height
	this.HeadNavX = 0;   	//  X pos
	this.HeadNavY = 0; 		// Y pos (above compass)

	this.HeadNavXC = this.HeadNavW / 2; // rel center X
	this.HeadNavYC = this.HeadNavH / 2; // rel center Y
	
	//
	this.offset     = 20; // offset of the Scale items
	this.scale      = 10; // the scale number steps
	this.nPad       = "   "; // number pad format # digits expected
}

// the GEO instance to use
const comp_geo = new Comp_geo();

// The verticalSpeed object to use outside 
//   as_obj.draw() will do the job
const comp_obj = {
	// local only - draw the scale of the compasss
	drawScale : function()
	{
		var ctx = gp_geo.Canvas.getContext("2d");

		//	we are at relative 0/0
        ctx.save();
			// set the origin to center
			ctx.translate(comp_geo.R, comp_geo.R); 

			// dashes at 45 deg
			ctx.strokeStyle = gpGUI.colFrame;
			ctx.fillStyle = gpGUI.colFrame;
			ctx.lineWidth = 2 ;
			ctx.beginPath();
			var i;
			for ( i=0; i<8; i++) {
				ctx.moveTo(0, -comp_geo.R+comp_geo.offset);
				ctx.lineTo(0, -comp_geo.R);
				ctx.rotate(Math.PI/4);
				// toggle the lineWith for major, minor
			}
			ctx.stroke();
			// 
			// Mark triangle			
			ctx.beginPath();
			ctx.moveTo(-10, -comp_geo.R);
			ctx.lineTo(0, -comp_geo.R+comp_geo.offset);
			ctx.lineTo(10, -comp_geo.R);
			ctx.closePath();
			ctx.fill();
			ctx.fill();

			// major
			ctx.lineWidth = 4;
			ctx.beginPath();
			for ( i=0; i<4; i++) {
				ctx.moveTo(0, -comp_geo.R+comp_geo.offset);
				ctx.lineTo(0, -comp_geo.R);
				ctx.rotate(Math.PI/2);
				// toggle the lineWith for major, minor
			}
			ctx.stroke();

			// Center Plane little drawing..
			ctx.lineWidth = 1 ;
			ctx.beginPath();
			ctx.moveTo(0, -20); 
			ctx.lineTo( -2, -18); ctx.lineTo( -4, -14); ctx.lineTo(-4, -6); ctx.lineTo(-20, 4); ctx.lineTo( -20, 8); ctx.lineTo( -4, 6); ctx.lineTo( -4, 14); ctx.lineTo(-8, 18); ctx.lineTo(-8, 20);
			ctx.lineTo(0, 20); 
			ctx.lineTo(8, 20); ctx.lineTo(8, 18); ctx.lineTo( 4, 14); ctx.lineTo( 4, 6); ctx.lineTo( 20, 8); ctx.lineTo(20, 4); ctx.lineTo(4, -6); ctx.lineTo( 4, -14); ctx.lineTo( 2, -18); 
			ctx.closePath();
			ctx.fill();

			// Track Heading scale -18..18°
			ctx.lineWidth = 2 ;
			ctx.rotate(-18*Math.PI/180);
			ctx.beginPath();
			for ( i=-2; i<3; i++) {
				ctx.moveTo(0, -comp_geo.R+comp_geo.offset);
				ctx.lineTo(0, -comp_geo.R);
				ctx.rotate(9*Math.PI/180);
			}
			ctx.stroke();
			
			
        ctx.restore();
	},

	// draw the compass rose
	draw_compass : function (THEAD)
	{
		if ( THEAD == null) return; // dont draw compass if not available

		var ctx = gp_geo.Canvas.getContext("2d");

		//	we are at relative 0/0
        ctx.save();
			// set the origin to center
			ctx.translate(comp_geo.R, comp_geo.R); 

			// inner area
			ctx.fillStyle = gpGUI.colCompass;
			ctx.globalAlpha=0.4; // draw transparent rectangle frame
			ctx.beginPath();
			ctx.moveTo(0, 0);
			ctx.arc(0,0,comp_geo.R-comp_geo.offset,0,2*Math.PI);
			ctx.moveTo(0, 0);
			ctx.arc(0,0,comp_geo.R,(-90-24)*Math.PI/180, (-90+24)*Math.PI/180 );
			ctx.fill();
			ctx.globalAlpha=1.0;

			// Set to heading
			ctx.rotate(-THEAD*Math.PI/180);
			
			// draw inner scale and labels
			ctx.font = gpGUI.font24s;
			ctx.fillStyle = gpGUI.colSmoke; // almost white
			ctx.textAlign = "center";
			ctx.textBaseline="top";
		
			ctx.strokeStyle = gpGUI.colSmoke;
			ctx.lineWidth=2;
			var y = -comp_geo.R+comp_geo.offset;
			ctx.beginPath();
			var i;
			for ( i=0; i<72; i++) {
				ctx.moveTo(0, y);
				if (( i % 2)==0){
					ctx.lineTo(0, y+comp_geo.offset);
					switch (i){
						case 0:
							ctx.fillText("N", 0, y+comp_geo.offset);
						break;
						case 18:
							ctx.fillText("E", 0, y+comp_geo.offset);
						break;
						case 36:
							ctx.fillText("S", 0, y+comp_geo.offset);
						break;
						case 54:
							ctx.fillText("W", 0, y+comp_geo.offset);
						break;
						case 6:
							ctx.fillText("3", 0, y+comp_geo.offset);
						break;
						case 12:
							ctx.fillText("6", 0, y+comp_geo.offset);
						break;
						case 24:
							ctx.fillText("12", 0, y+comp_geo.offset);
						break;
						case 30:
							ctx.fillText("15", 0, y+comp_geo.offset);
						break;
						case 42:
							ctx.fillText("21", 0, y+comp_geo.offset);
						break;
						case 48:
							ctx.fillText("24", 0, y+comp_geo.offset);
						break;
						case 60:
							ctx.fillText("30", 0, y+comp_geo.offset);
						break;
						case 66:
							ctx.fillText("33", 0, y+comp_geo.offset);
						break;
					}
				}
				else {
					ctx.lineTo(0, y+comp_geo.offset/2);
				}
				ctx.rotate(Math.PI/36.0);
				// toggle the lineWith for major, minor
			}
			// draw it
			ctx.stroke();
			
			// leave with the origin reset to the canvas origin
        ctx.restore();
	},

	// print Heading and Track
	draw_heading : function(THEAD)
	{
		if ( THEAD != null) {
			THEAD = Math.round(THEAD);
		}
	
		var ctx = gp_geo.Canvas.getContext("2d");
		// set the origin to pos of HEAD field and assume a relative drawing 
        ctx.save();
			ctx.translate(comp_geo.HeadX, comp_geo.HeadY); 
			// frame and background		
			ctx.fillStyle = gpGUI.colBGLabel;
			ctx.strokeStyle = gpGUI.colFrame;
			ctx.fillRect(0,0, comp_geo.HeadW, comp_geo.HeadH);
			// for the rest clip the usable area
			ctx.rect(1,1, comp_geo.HeadW-2, comp_geo.HeadH-2);
			ctx.clip();
			// finally the indicated number
			ctx.globalAlpha=1.0; // opaque
			ctx.font = gpGUI.font32;
			ctx.fillStyle = gpGUI.colWhite;
			ctx.textAlign = "center";
			ctx.textBaseline="middle";
			if ( THEAD != null ) {
				ctx.fillText( THEAD.toString() + "°", comp_geo.HeadXC, comp_geo.HeadYC); // middle align
			} 
			else {
				ctx.fillText( "- - -", comp_geo.HeadXC, comp_geo.HeadYC); // middle align
			}
		// leave with the origin reset to the canvas origin
        ctx.restore();
	},
	

	// print Track - give TRACK and HEADING
	draw_track : function(TRACK, HEADING)
	{
		if ( TRACK == null) return;
		if ( HEADING == null) return;
		
		var ctx = gp_geo.Canvas.getContext("2d");
		// set the origin to pos of HEAD field and assume a relative drawing 
        ctx.save();
			ctx.translate(comp_geo.R, comp_geo.R); 
			// Set to track offset vs heading
			ctx.rotate(-(HEADING-TRACK)*Math.PI/180);

			ctx.fillStyle = gpGUI.colPurple;
			var y = -comp_geo.R + 2*comp_geo.offset; // middle base of indicator
			ctx.beginPath();
			ctx.moveTo(0, y); 
			ctx.lineTo(-5, y-comp_geo.offset/2);
			ctx.lineTo(0, y-comp_geo.offset);
			ctx.lineTo(+5, y-comp_geo.offset/2);
			ctx.closePath();
			ctx.fill();
        ctx.restore();
	},
	

	// print Track - give TRACK and HEADING
	draw_headingNavBug : function(NAVHEAD, HEADING)
	{
		if ( NAVHEAD == null) return;
		if ( HEADING == null) return; // don't show navBug if heading is not avail
		
		var ctx = gp_geo.Canvas.getContext("2d");
        ctx.save();
			ctx.translate(comp_geo.R, comp_geo.R); 
			// Set to Nav Heading
			ctx.rotate(-(HEADING-NAVHEAD)*Math.PI/180);

			ctx.fillStyle = gpGUI.colNav;
			var y = -comp_geo.R + comp_geo.offset; // middle base of indicator
			ctx.beginPath();
			ctx.moveTo(0, y+10); 
			ctx.lineTo(-8, y); ctx.lineTo(-15, y+2); ctx.lineTo(-15, y+20); ctx.lineTo(-13, y+18); 
			ctx.lineTo(+13, y+18); ctx.lineTo(+15, y+20); ctx.lineTo(+15, y+2); ctx.lineTo(+8, y); 
			ctx.closePath();
			ctx.fill();
        ctx.restore();
	},
	

	// print Nav Heading
	draw_headingNav : function(NAVHEAD)
	{
		if ( NAVHEAD == null) return;
		NAVHEAD = Math.round(NAVHEAD);

		var ctx = gp_geo.Canvas.getContext("2d");
		// set the origin to pos of HEAD field and assume a relative drawing 
        ctx.save();
			ctx.translate(comp_geo.HeadNavX, comp_geo.HeadNavY); 
			// frame and background		
			ctx.fillStyle = gpGUI.colBGLabel;
			ctx.strokeStyle = gpGUI.colFrame;
			ctx.fillRect(0,0, comp_geo.HeadNavW, comp_geo.HeadNavH);
			ctx.strokeRect(0,0, comp_geo.HeadNavW, comp_geo.HeadNavH);
			// for the rest clip the usable area
			ctx.rect(1,1, comp_geo.HeadNavW-2, comp_geo.HeadNavH-2);
			ctx.clip();
			// finally the indicated number
			ctx.globalAlpha=1.0; // opaque
			ctx.font = gpGUI.font24;
			ctx.fillStyle = gpGUI.colNav;
			ctx.textAlign = "center";
			ctx.textBaseline="middle";
			ctx.fillText( "HDG " + NAVHEAD.toString() + "°", comp_geo.HeadNavXC, comp_geo.HeadNavYC); // middle align
		// leave with the origin reset to the canvas origin
        ctx.restore();
	},
	

	// print TrackIndicator 
	draw_trackIndicator : function(TRACKRATE)
	{
		if ( TRACKRATE == null) return;
		if ( Math.abs(TRACKRATE) < 0.05 ) return; // don't show if small
		var trate = ( Math.abs(TRACKRATE)>4.0 ) ? 4.0 : TRACKRATE; // limit, some seem to just report crap... i.e. above 4 r is killed off		
		
		var ctx = gp_geo.Canvas.getContext("2d");
		// set the origin to pos of HEAD field and assume a relative drawing 
        ctx.save();
			// set the origin to center
			ctx.translate(comp_geo.R, comp_geo.R); 
			ctx.strokeStyle = gpGUI.colPurple;
			ctx.lineWidth = 3;
			var trkDeg = trate * 18; //deg for now
			// Track Heading scale -18..18°
			ctx.beginPath();
			if (TRACKRATE<0) {
				ctx.arc(0,0,comp_geo.R-comp_geo.offset+4,-90*Math.PI/180, (-90+trkDeg)*Math.PI/180, true ); // left of north
			} else {
				ctx.arc(0,0,comp_geo.R-comp_geo.offset+4,-90*Math.PI/180, (-90+trkDeg)*Math.PI/180 );
			}
			ctx.stroke();
        ctx.restore();
	},
	

	// MAIN call to draw the airspeed instrument
	draw : function(THEAD, NHEAD, TRACK, TRACKRATE) 
	{
		var ctx = gp_geo.Canvas.getContext("2d");

		var value = THEAD; // set to uniform the routines
//		value = CenterValue; // DEVELOP we take it from global until having a real number

        ctx.save();
			// set the origin to pos of AS indicator and assume a relative drawing 
			ctx.translate(comp_geo.X, comp_geo.Y); 
	
			// draw the compass 
			this.draw_compass(THEAD);
			// makes the scale indicators
			this.drawScale();
			this.draw_headingNavBug(NHEAD, THEAD);
			this.draw_track(TRACK, THEAD); //TRACK, HEADING
	
			this.draw_heading(THEAD);
			this.draw_headingNav(NHEAD);
			this.draw_trackIndicator(TRACKRATE);
		// leave with the origin reset to the canvas origin
        ctx.restore();
		
		// TESTING ONLY - incr by clicking
		CenterValue+=1;
	}
}





